tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts

```diff
@@ -308,8 +311,27 @@ export function buildOrReturnStateWithPinnedCopy(
     throw new Error('Cannot pin a card without metadata');
   }
   nextCardMetadataMap[pinnedCardId] = metadata;
+
+  // If the card has state
+  // 1) Create a deep copy of the state
+  // 2) Set the new state equal to the copy
+  // 3) Set any pinned versions equal to the copy
   if (nextCardStateMap[cardId]) {
-    nextCardStateMap[pinnedCardId] = {...cardStateMap[cardId]};
+    const nextCardState = Object.entries(nextCardStateMap[cardId]).reduce(
+      (cardState, [key, value]) => {
+        if (Array.isArray(value)) {
+          cardState[key as keyof CardState] = [...value] as any;
+        } else if (typeof value === 'object') {
+          cardState[key as keyof CardState] = {...value} as any;
+        } else {
+          cardState[key as keyof CardState] = value as any;
+        }
+        return cardState;
+      },
+      {} as Partial<CardState>
+    );
```

Comment:
I am worried about the pinned card state pointing to the same object as the original card. I'm not sure how the NgRx deep freeze you mentioned works so I'm hesitant to say it would resolve the issue.

I would rather developers not have to worry about whether newly added properties are copied over when a card is pinned.

I'm really not sure why I didn't think making a true deep copy was a good idea, the resulting code is much less ledgible and significantly more fragile with the only benefit being a slight performance boost. FWIW `structuredClone` is a slightly more performant solution than `JSON.parse(JSON.stringify())` but it's only supported by new browser versions so I'm intentionally not using it here.

Reply:
Here is a simple example of card entanglement via a shared state reference
1) The user expands the card.
`card1State = {tableExpanded: true}`
The card state map is now `{ card1: card1State}`
2) The user pins the card.
The card state map is now `{card1: card1State, pinnedCard1: card1State}`
3) The user collapses the pinned cards table
The card state map "unchanged"  `{card1: card1State, pinnedCard1: card1State}`
Except that `card1State = {tableExpanded: false}` thus collapsing the table for both cards.

Reply:
I don't mind the type assertion in this case. The object was initially `Partial<CardState>`, after being serialized and deserialized will still be `Partial<CardState>`

Reply:
For the case you illustrate, I don't think NgRx will allow that to happen.

You can't do this, for example:
```
state.cardStateMap[pinnedCardId].tableExpanded = false
```

That will silently fail because of the frozen state. The tableExpanded property will continue to be true and neither card will collapse.

NgRx is basically forcing you to do something like this:

```
state.cardStateMap[pinnedCardId] = {
  ...state.cardStateMap[pinnedCardId],
  tableExpanded: false
};
```

That will work around the frozen state AND will decouple the pinned card's state from the original card's state. Pinned card will be collapsed but original card will be expanded.

Reply:
So to paraphrase, it is okay if both cards reference the same object temporarily because when an attribute changes the object will need to be reconstructed anyway. That makes sense, I guess I was overly worried about that intermediate state.